KAFKA-7333; Protocol changes for KIP-320

hachikuji opened the pull request:
This patch contains the protocol updates needed for KIP-320 as well as some of the basic consumer APIs (e.g. `OffsetAndMetadata` and `ConsumerRecord`). The inter-broker format version has not been changed and the brokers will continue to use the current API versions.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

hachikuji committed:
KAFKA-7333; Protocol changes for KIP-320

hachikuji commented:
@lindong28 Would you have time to review this? The changes are mostly mechanical. I have avoided updating any of the internal handling logic other than in trivial ways. Basically I'm trying to get all of the boilerplate out of the way so that reviewing the changes to the logic will be easier.

The test failure is because the replica fetcher is (intentionally) not using the latest version of the fetch api. I'm thinking of ways to fix this, but one option is to not expose the new versions in the corresponding `XXXRequest.schemaVersions()` methods until the new behavior has been implemented.

lindong28 commented:
@hachikuji Sure, I will review it this week.

hachikuji committed:
Prefer use of optional integer as leader epoch return type

hachikuji committed:
Minor cleanups

lindong28 commented:
Hey @hachikuji, since the patch is still being updated, could you let me know when it is ready for review :)

hachikuji commented:
@lindong28 Yeah, sorry about that. I pushed a few minor tweaks as I was going over the code myself. Should be ready early next week, but I'll let you know.

lindong28 commented:
@hachikuji No problem! Yeah let me know when it is ready.

hachikuji committed:
Bump inter-broker version and update ReplicaFetcher

hachikuji committed:
Fix rebase breakage

hachikuji commented:
@lindong28 This should be ready for review when you have a chance. I have bumped the inter-broker api version. Until the logic for KIP-320 is merged, the replica fetcher will use the updated protocols, but will retain the current behavior. Does that seem reasonable?

hachikuji committed:
Update failing tests for new api versions

lindong28 commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -168,10 +170,12 @@ public Fetcher(LogContext logContext,
+        final Optional<Integer> leaderEpoch; // null if leader epoch is not known
 
-        OffsetData(long offset, Long timestamp) {
+        OffsetData(long offset, Long timestamp, Optional<Integer> leaderEpoch) {
```
Since `OffsetData()` is not exposed to user, can we just use `int leaderEpoch` as the type of the constructor parameter and private variable? Maybe we can consistently use `RecordBatch.NO_PARTITION_LEADER_EPOCH` to indicate missing leaderEpoch for all non-user-facing code so as to optimize performance.

lindong28 commented on clients/src/main/java/org/apache/kafka/clients/consumer/OffsetAndMetadata.java:
```diff
@@ -19,23 +19,39 @@
+    // We use null to represent the absence of a leader epoch to simplify serialization.
+    // I.e., older serializations of this class which do not have this field will automatically
+    // initialize its value to null.
+    private final Integer leaderEpoch;
```
The current patch uses `Integer leaderEpoch = null` to indicate missing leader epoch. An alternative way is to use `int leaderEpoch = RecordBatch.NO_PARTITION_LEADER_EPOCH`. Since `RecordBatch.NO_PARTITION_LEADER_EPOCH` is used in many other places, I am wondering if it is a bit more consistent to also use `int leaderEpoch = RecordBatch.NO_PARTITION_LEADER_EPOCH` here and in `OffsetAndTimestamp`.

hachikuji replied:
That was what I attempted to do initially. Unfortunately we cannot depend on initializers during object deserialization. When deserializing old versions, the new leader epoch field will be initialized to the default value for the corresponding type. So if we use `int leaderEpoch = RecordBatch.NO_PARTITION_LEADER_EPOCH`, then the value after deserialization will actually be 0. Since 0 is a valid value for this field, I couldn't think of a way to make it work. On the other hand, the default value for `Integer` is null, so it works. 

I really wish we hadn't made this class `Serializable`, but tough to take it away now. One possibility is to make this an exceptional case and to prefer the use of the `int` type for other classes.

ijuma replied:
@hachikuji You could probably use `readObject` if you want right?

hachikuji replied:
I don't think so. At least I could not find a way to make it work. If you override `readObject`, you still have to call `defaultReadObject`, but this will initialize all of the fields. But then how do you know whether the initialized value was read from the serialized data or is simply the default value?

hachikuji replied:
Discussed briefly offline, but the messy alternative is read each field individually (following the specification) and either change to non-final fields or do some tricky reflection to initialize the fields. Seems simpler to leave it as is, but let me know me if either of you disagree.

lindong28 replied:
Ah I see. I agree it is simpler to leave it as is.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -36,153 +37,162 @@
+    private static final Field.Int32 REPLICA_ID = new Field.Int32("replica_id",
+            "Broker id of the follower. For normal consumers, use -1.");
+    private static final Field.Int64 FETCH_OFFSET = new Field.Int64("fetch_offset", "Message offset.");
+    private static final Field.Int32 PARTITION_MAX_BYTES = new Field.Int32("partition_max_bytes",
```
Here the patch changes the name of an existing field in FetchRequest V0 from `max_bytes` to `partition_max_bytes`. Would it cause confusion for developers who implement their custom client library? And do we need to file a KIP to change the name of the request schema field?

hachikuji replied:
Hmm.. the names are only used internally. Can you clarify how it would impact client developers to change this name?

lindong28 replied:
Not sure if it is completely internal. It seems that field name is part of the request specification and request specification is public API. My understanding is that we need to have a KIP for such public API change.

More specifically, custom client library or tool (e.g. https://github.com/linkedin/Burrow) needs to specify the serialization/deserialization of kafka request/response. It seems that developers of these libraries may use kafka source code or documentation as source of truth of the request/response specification. I am a bit concerned that these developers will be confused if we change the name of old schema. Does this make sense?

hachikuji replied:
We have not previously considered these field names to be part of the public API, but it's a reasonable question whether we should. Just to be clear, there is no direct dependence on the names in tools like burrow, but your concern is that changing the names may cause confusion? I can appreciate that. This particular name change is probably not strictly needed, but also the potential for confusion either way is probably minimal (it emphasizes but does not change the usage). In the case below, I renamed `epoch` to `session_epoch` because the fetch request now includes the leader epoch as well. I think the potential for confusion is greater if we leave it as is.

lindong28 replied:
Yeah I totally agree `partition_max_bytes` is a better name than `max_bytes`. I am wondering if it is better to change the field name for FetchRequest V9 only and keep the name unchanged for FetchRequest V8 and earlier. The benefit of this approach is that developers (both Kafka developer and other library developer) can keep track of the evaluation of the schema (both its structure and field name change). Are you concerned that this would cause confusion if the field name is different for different versions of the schema in `FetchRequest.java`?

I was thinking that field name is part of API because field name is used by developer to identify the request/response schema even though field name is not used by Kafka server code to identify the schema. I may be overly worried. It will be great if we can have input from one more developer to comment on this. What do you think?

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -36,153 +37,162 @@
-    public static final Field.Int32 EPOCH = new Field.Int32("epoch", "The fetch epoch");
+    // V7 added incremental fetch requests.
+    private static final Field.Int32 SESSION_ID = new Field.Int32("session_id", "The fetch session ID");
+    private static final Field.Int32 SESSION_EPOCH = new Field.Int32("session_epoch", "The fetch session epoch");
```
Here the patch changes the field name from `session_epoch` to `epoch`. I am wondering if it would cause confusion to developer of custom client library and whether we need to file a KIP for the name change.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -36,153 +37,162 @@
+    // V9 adds the current leader epoch (see KIP-320)
+    private static final Schema FETCH_REQUEST_PARTITION_V9 = new Schema(
+            PARTITION_ID,
+            CURRENT_LEADER_EPOCH,
```
Can we add the `CURRENT_LEADER_EPOCH` right after the `LOG_START_OFFSET` field to exactly follow the specification in KIP-320?

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -207,21 +217,33 @@
+            this.currentLeaderEpoch = currentLeaderEpoch;
+        }
+
+        public Optional<Integer> currentLeaderEpoch() {
```
Since `PartitionData.currentLeaderEpoch()` is not exposed to kafka user, would it be faster and simpler to just return `int` instead of `Optional<Integer>` as the leaderEpoch?

hachikuji replied:
I was a little torn on this. We have a bad habit of misusing sentinel values internally. For example, see https://github.com/apache/kafka/commit/a5318722c74ec0279a2e831e93e4164d9adcb8a1#diff-25ea40d139b41c948888c2a4f13c9e82. One option might be to drop this API and instead expose a method for the fencing check. Alternatively we can return the raw type and just try to be careful using it.

ijuma replied:
Yes, sentinels are terrible. Even worse than `null`, which is saying something. The cost of us incorrectly writing -1 as the watermark in the checkpoint files has been immense. Probably months of developer/support time to deal with it.

lindong28 replied:
Sounds good. I am good with this approach if you and Ismael prefer Optional<Integer>.

ijuma replied:
Btw, to clarify, I think it's a totally reasonable question given that we use sentinels a lot. I have seen a lot of costly bugs due to them, so I think we should avoid them as much as we can.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java:
```diff
@@ -43,13 +43,20 @@
+ *
+ * Possible error codes:
+ *
+ *  OFFSET_OUT_OF_RANGE (1)
```
Can we also `FENCED_LEADER_EPOCH` and `UNKNOWN_LEADER_EPOCH` for relevant responses such as `FetchResponse` and `ListOffsetResponse`?

hachikuji replied:
I was planning to add this when I introduced the new error codes. Shall we do that in this patch?

lindong28 replied:
Sounds good. This can be added later.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java:
```diff
@@ -85,42 +90,47 @@
-            new Field(PRODUCER_ID_KEY_NAME, INT64, "The producer id associated with the aborted transactions"),
-            new Field(FIRST_OFFSET_KEY_NAME, INT64, "The first offset in the aborted transaction"));
+    // aborted transaction field names
+    private static final Field.Int64 LAST_STABLE_OFFSET = new Field.Int64("last_stable_offset",
```
Just my personal opinion. It seems that the original way of specifying all possible fields at the beginning of the class is more readable than putting these names closer to the first version of the schema that uses the field.

There is similar discussion in https://github.com/apache/kafka/pull/5491#issuecomment-413283121. The situation is not exactly the same. I am not strong on this. Maybe @ijuma can also comment on which style is preferred.

hachikuji replied:
I agree it's subjective. I thought it was useful to initialize the field along with the schema version that introduced it. I think it's a bit nasty no matter what to tell you the truth. I hope eventually we're not hand-coding these schemas anymore.

lindong28 replied:
I am not very strong on this. On the other hand I don't feel strong about committing this refactor change based on one person's opinion. @ijuma Could you comment which style is preferred?

hachikuji replied:
Yeah, I don't have a strong preference either. The main thing I was trying to accomplish is avoiding the duplication of the type and the documentation string. Where we put the field is less of a concern.

ijuma replied:
I'd say that defining the fields near its schema provides a bit more context once the request/response becomes complex. Once you add multiple schema versions, it becomes more nuanced and it's unclear which is better. Either way works for me. It would be nice to auto-generate this as @hachikuji says.

lindong28 replied:
So it looks like there is no clearly better choice. Typically I feel it is probably better to refactor code only if the new style is better. The reason is mainly to reduce the code conflict with patches from other contributors and also to reduce the need for other contributors to learn the new pattern.

Since @ijuma seems OK with the new style, we can do the refactor as is.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/ListOffsetResponse.java:
```diff
@@ -47,57 +58,68 @@
 
+    // V4 bumped for the addition of the current leader epoch in the request schema and the
+    // leader epoch in the response partition data
+    private static final Schema LIST_OFFSET_RESPONSE_PARTITION_V4 = new Schema(
```
Hmm.. it seems that the order of schema fields specified in KIP-320 is different from the existing code. It maybe inconvenience for developer of Kafka and custom client library to reference the KIP-320 and understand the protocol change. Can KIP-320 be updated to show the exact request/response schema?

hachikuji replied:
Ack. I've updated KIP-320. I tried to use the following order consistently: partition, leader epoch, offset (if present).

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java:
```diff
@@ -35,65 +36,55 @@
             ERROR_CODE,
             PARTITION_ID,
-            new Field(LEADER_KEY_NAME, INT32, "The id of the broker acting as leader for this partition."),
+            LEADER,
```
Some fields (e.g. LEADER) are transformed into e.g. `Field.Int32` whereas other fields stays as e.g. `new Field(REPLICAS_KEY_NAME, new ArrayOf(INT32), ...)`. Due to this mixture the new style is less consistent than the previous style. If the goal of the refactor is to make the code more readable, can we also define e.g. Field.Array(Field field, String name, String doc...) to make the code further easier to read?

hachikuji replied:
The goal was to avoid type/doc duplication, but I agree that the result can be hard on the eyes. Let me give your suggestion a try.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java:
```diff
@@ -354,14 +373,18 @@ public Cluster cluster() {
                     internalTopics.add(metadata.topic);
-                for (PartitionMetadata partitionMetadata : metadata.partitionMetadata)
+                for (PartitionMetadata partitionMetadata : metadata.partitionMetadata) {
+                    Integer leaderEpoch = partitionMetadata.leaderEpoch == RecordBatch.NO_PARTITION_LEADER_EPOCH ?
```
There are a few other places where `int leaderEpoch = RecordBatch.NO_PARTITION_LEADER_EPOCH` is used to indicate missing leader epoch. Can we consistently use the same approach?

hachikuji replied:
Yeah, sorry about that. I was struggling to find a consistent pattern given the complication of `OffsetAndMetadata` and the apparently frowned-upon use of `Optional` in fields. It is helpful generally when argument types match return types since it makes composition easier, but I'm not sure how you can achieve that with the standard usage advice for `Optional`. I think our options are the following:

1. Use `Optional<Integer>` everywhere (including in fields) except within `OffsetAndMetadata`.
2. Use nullable integer internally, but expose `Optional<Integer>` from public APIs.
3. Use primitive int with a sentinel internally, but expose `Optional<Integer>` from public APIs.

I was trying hard to avoid the third option because unsafe use of sentinel values can cause significantly more damage than unsafe use of an optional or nullable value. At least in the latter case, you get a fairly clear exception which is raised from the point of usage. Improper use of a sentinel value can have a massive impact at a distance from the usage (as mentioned above).

My inclination at the moment is probably option 1, but I'd like to hear other opinions.

ijuma replied:
I prefer `1` which is the approach we use in Scala and I believe has worked out pretty well. It's annoying that Java came up with its own weird guidelines about `Optional` usage. One concrete outcome is that it doesn't implement `Serializable`. Since we only implement `Serializable` in a couple of classes, this doesn't seem like a blocker.

lindong28 replied:
Sounds good. Let's use option 1.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetCommitRequest.java:
```diff
@@ -17,109 +17,134 @@
+
+    private static final Schema OFFSET_COMMIT_REQUEST_PARTITION_V6 = new Schema(
+            PARTITION_ID,
+            LEADER_EPOCH,
```
It seems that the position of the new field `LEADER_EPOCH` is different from that is specified in the KIP-320. And KIP-320 specified `LastLeaderEpoch` in OffsetCommitRequest and OffsetFetchResponse. Can we update the KIP as appropriate and also make the code consistent with schema specification in the KIP?

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchResponse.java:
```diff
@@ -64,47 +78,56 @@
+
+    private static final Schema OFFSET_FETCH_RESPONSE_PARTITION_V5 = new Schema(
+            PARTITION_ID,
+            LEADER_EPOCH,
```
It seems that the position of the new field `LEADER_EPOCH` is different from that is specified in the KIP-320. And KIP-320 specified `Timestamp` and `LastLeaderEpoch` in OffsetFetchResponse. Can we update the KIP as appropriate and also make the code consistent with schema specification in the KIP?

lindong28 commented on core/src/main/scala/kafka/api/ApiVersion.scala:
```diff
@@ -75,7 +75,9 @@ object ApiVersion {
     // Introduced new schemas for group offset (v2) and group metadata (v2) (KIP-211)
-    KAFKA_2_1_IV0
+    KAFKA_2_1_IV0,
+    // New Fetch, OffsetsForLeaderEpoch, and ListOffsets schemas (KIP-320)
```
From the existing protocol version comment, it is not clear what is the information required for each protocol version bump. I am wondering whether it is useful to list all the requests that are changed in KIP-320 here.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetsForLeaderEpochRequest.java:
```diff
@@ -158,4 +181,21 @@ public AbstractResponse getErrorResponse(int throttleTimeMs, Throwable e) {
+
+    public static class PartitionData {
+        private final int currentLeaderEpoch;
+        public final int searchLeaderEpoch;
```
`searchLeaderEpoch` is different from field name `leaderEpoch` in the schema. In general it is probably preferred to be consistent. If we feel that searchLeaderEpoch is more intuitive and readable than leaderEpoch, would it make sense to just use this name in the schema in KIP-320?

hachikuji replied:
I was not too happy with the name `searchLeaderEpoch`, so I'm happy to revert. I was trying to avoid confusion, but perhaps it is clear enough with the current leader epoch distinguished.

lindong28 reviewed (commented):
Thanks for the patch @hachikuji!

I left some comments. Most comments are related to either the code style in refactor or the consistency in schema specification between this patch and KIP-320.

hachikuji commented:
@lindong28 Thanks a lot for the comments. I left a few responses. I'll wait to make sure we're on the same page before pushing updates.

lindong28 commented:
@hachikuji Thanks for the explanation. The explanation makes sense. I have replied to all your latest comments.

hachikuji committed:
Use Optional consistently to represent leader epoch

hachikuji committed:
Revert name change in OffsetsForLeaderEpochRequest

hachikuji committed:
Fix rebase breakage

hachikuji committed:
TxnOffsetCommit needs update as well

hachikuji committed:
Support for array field types

hachikuji committed:
fix checkstyle errors

hachikuji commented:
@lindong28 This is probably ready for another look when you have a chance. I have fixed the inconsistent usage of `leaderEpoch` and I have refactored the schemas (following your suggestion) to improve readability. 

There are a couple things that we missed in the KIP that are included in this patch. First, we needed to update the `TxnOffsetCommit` API as well to include the leader epoch. Second, I've added the leader epoch to `PartitionInfo` which is returned from the `partitionsFor` API in the producer and the consumer. If you think the changes are reasonable, I will update the KIP and sent a note to the KIP thread.

lindong28 commented:
@hachikuji Thanks for the update! Certainly, I will review the patch and let you know soon.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/ListOffsetRequest.java:
```diff
@@ -187,25 +179,34 @@ public String toString() {
+            this(timestamp, maxNumOffsets, Optional.empty());
+        }
+
+        public static PartitionData withCurrentLeaderEpoch(long timestamp, Optional<Integer> currentLeaderEpoch) {
```
`withCurrentLeaderEpoch(...)` is essentially the same as the following constructor for `PartitionData`. Could you explain a bit why we use a static method instead of constructor?

```
        public PartitionData(long timestamp, Optional<Integer> currentLeaderEpoch) {
            this(timestamp, 1, currentLeaderEpoch);
        }
```

hachikuji replied:
Ack. Now that that types no longer conflict, we can use a normal constructor.

lindong28 commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -698,9 +704,10 @@ public void onFailure(RuntimeException e) {
-    private RequestFuture<ListOffsetResult> sendListOffsetRequest(final Node node,
-                                                                  final Map<TopicPartition, Long> timestampsToSearch,
-                                                                  boolean requireTimestamp) {
+    private RequestFuture<ListOffsetResult> sendListOffsetRequest(
```
nits: Most methods in this class use the previous style of indentation. Could we make the indentation style consistent?

lindong28 commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -756,15 +763,16 @@ private void handleListOffsetResponse(Map<TopicPartition, Long> timestampsToSear
                             topicPartition, offset);
                     if (offset != ListOffsetResponse.UNKNOWN_OFFSET) {
-                        OffsetData offsetData = new OffsetData(offset, null);
+                        OffsetData offsetData = new OffsetData(offset, null, null);
```
Should it be `new OffsetData(offset, null, Optional.empty())`?

lindong28 commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecord.java:
```diff
@@ -225,6 +262,15 @@ public int serializedValueSize() {
+    public Optional<Integer> leaderEpoch() {
+        return leaderEpoch;
+    }
+
```
Can we also update toString() to include `leaderEpoch`?

lindong28 commented on clients/src/main/java/org/apache/kafka/clients/consumer/OffsetAndTimestamp.java:
```diff
@@ -16,20 +16,31 @@
 public final class OffsetAndTimestamp {
     private final long timestamp;
     private final long offset;
+    private final Integer leaderEpoch;
```
We probably want to only `Integer leaderEpoch` in `OffsetAndMetadata` to simplify its serialization. Since `OffsetAndTimestamp` does not implement `Serializable`, can we make it `Optional<Integer> leaderEpoch` for consistency?

hachikuji replied:
Ack. Missed this one in the previous update.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/protocol/CommonFields.java:
```diff
@@ -27,7 +27,12 @@
+    public static final Field.Int32 CURRENT_LEADER_EPOCH = new Field.Int32("current_leader_epoch",
+            "The current leader epoch, if provided, is used to fence consumers/replicas with old metadata. " +
+                    "If the provided value does not match the epoch of the current leader, then the broker " +
+                    "will respond with either UNKNOWN_LEADER_EPOCH (if the epoch is larger) " +
```
Would it be better to replace `if the epoch is larger` with `if the current_leader_epoch is larger`, and replace `if the epoch is smaller` with `if the current_leader_epoch is smaller`?

hachikuji replied:
Ack. I will try to rephrase.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/ListOffsetResponse.java:
```diff
@@ -31,73 +30,98 @@
+    // V4 bumped for the addition of the current leader epoch in the request schema and the
+    // leader epoch in the response partition data
+    private static final Field PARTITIONS_V4 = PARTITIONS.withFields(
+            PARTITION_ID,
```
Could you update the `ListOffsetResponse` schema in KIP-320 to make it consistent with the code?

hachikuji replied:
Thanks. I noticed a couple additional missing fields in the documented schema (e.g. throttle time). I've updated the KIP. I will push the changes shortly. I've settled on the following order:

1. Current leader epoch always follows partition id in requests
2. When leader epoch is relative to an offset, then it is listed after the offset. This includes the offset commit apis and list offsets.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java:
```diff
@@ -35,143 +34,171 @@
+    // partition level fields
+    private static final Field.Int32 LEADER = new Field.Int32("leader",
+            "The id of the broker acting as leader for this partition.");
+    private static final Field.Int32 LEADER_EPOCH = new Field.Int32("leader_epoch",
```
Should this be replaced by `CommonFields.LEADER_EPOCH`? If not, say there is different in semantics, can we make the difference more explicit in field name and/or doc?

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetCommitRequest.java:
```diff
@@ -31,95 +29,112 @@
+
+    // V6 adds the leader epoch to the partition data
+    private static final Field PARTITIONS_V6 = PARTITIONS.withFields(
+            PARTITION_ID,
```
Could we update the schema for the OffsetCommitRequesst in KIP-320 so that its partition level fields is consistent with the code, i.e. PARTITION_ID, COMMITTED_LEADER_EPOCH. COMMITTED_OFFSET and COMMITTED_METADATA.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetCommitRequest.java:
```diff
@@ -191,18 +213,12 @@ public Builder setGenerationId(int generationId) {
-                    return new OffsetCommitRequest(groupId, generationId, memberId, DEFAULT_RETENTION_TIME, offsetData, version);
-                default:
-                    throw new UnsupportedVersionException("Unsupported version " + version);
+            if (version == 0) {
```
It seems that the previous approach is a bit better and stricter because it throws `UnsupportedVersionException` for versions that are not supported. Also it may be good to use the same approach as build() for other requests for consistency.

hachikuji replied:
There is some inconsistency between request types. This is the only request I found that had all versions explicitly listed. I think we can probably move the version check into `AbstractRequest` to make the code friendlier to maintain while still ensuring the strict version validation.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchResponse.java:
```diff
@@ -32,79 +31,94 @@
     private static final Schema OFFSET_FETCH_RESPONSE_V4 = OFFSET_FETCH_RESPONSE_V3;
 
+    // V5 adds the leader epoch to the committed offset
+    private static final Field PARTITIONS_V5 = PARTITIONS.withFields(
```
Can we update the KIP-320 so that its schema for OffsetFetchResponse is consistent with the code in terms of both field order and field name (i.e. LastLeaderEpoch vs leaderEpoch)?

lindong28 commented on core/src/main/scala/kafka/api/ApiVersion.scala:
```diff
@@ -75,7 +75,9 @@ object ApiVersion {
     // Introduced new schemas for group offset (v2) and group metadata (v2) (KIP-211)
-    KAFKA_2_1_IV0
+    KAFKA_2_1_IV0,
+    // New Fetch, OffsetsForLeaderEpoch, and ListOffsets schemas (KIP-320)
```
nits: can we include Metadata, OffsetCommit, TxnOffsetCommit and OffsetFetch in the comment for completeness?

hachikuji replied:
I only mentioned the inter-broker protocols that were affected since we don't bump this version for client-only APIs. Does that make sense?

lindong28 replied:
Yeah this makes sense.

lindong28 commented on core/src/main/scala/kafka/api/ApiVersion.scala:
```diff
@@ -256,5 +258,12 @@ case object KAFKA_2_1_IV0 extends DefaultApiVersion {
   val shortVersion: String = "2.1"
   val subVersion = "IV0"
   val recordVersion = RecordVersion.V2
+  val id: Int = 17
```
I guess it is probably not a problem for most users and it is cleaner to make this change.

One question is, if user is using the trunk version where KAFKA_2_1_IV0 has id=18, and then they bump up to the trunk version where KAFKA_2_1_IV0 has id=17 and KAFKA_2_1_IV1 has id=18, would it cause problem?

hachikuji replied:
As far as I can tell, we only use the id for ordering internally, so I don't think this change has any impact.

lindong28 reviewed (commented):
@hachikuji Thanks for the update. The patch looks very good now. I left some minor comments.

I agree with the schema change for `TxnOffsetCommit`. Can you explain why we need to include leaderEpoch in `PartitionInfo`?

hachikuji commented:
@lindong28 Thanks for the review. I'll address your comments shortly. To answer your question, I think the change to `PartitionInfo` is not strictly necessary. I thought since we are exposing all the rest of the metadata and we expose leader epoch everywhere else, we may as well include it. Debugging is probably the only use case I can think of. Another option might be to expose it only in the AdminClient `TopicPartitionInfo`. We can also punt on it for now and add it later if we find a more compelling need. What do you think?

lindong28 commented:
@hachikuji In general I think it is probably good to add a new field to public API if we have concrete use-case for the field. My intuition is that we don't need it because leaderEpoch is probably useful in the client-side only when used with offset, which is currently not provided in `PartitionInfo`. Yeah let's punt on it for now and add it later if needed.

hachikuji commented:
@lindong28 Sounds good. I have updated the KIP with the `TxnOffsetCommit` schema. I'll send a note to the vote thread.

hachikuji commented:
@lindong28 Hmm.. On second thought, there was a reason why I wanted to include the leader epoch in `PartitionInfo`. We use this object inside `common.Cluster`, which is depended upon by `clients.Metadata`. If we don't add leader epoch to `PartitionInfo`. Then we'll need to propagate the information through an additional argument (probably a map) to `Metadata.update`. It's a bit annoying and unnatural, but we can do it if we want to keep leader epoch hidden. Perhaps we need to introduce an intermediate object to remove our internal dependence on a publicly exposed object.

lindong28 commented:
@hachikuji It seems that the common.Cluster in this PR currently does not have a field for `epoch`. So I probably don't understand why it is needed. I guess in general public API is more important than internal stuff. So it is probably preferred to keep the public API concise and add complexity in internal implementation instead.

hachikuji commented:
Yeah, that's my point. When we implement the rest of KIP-320, we'll need to propagate the leader epoch from the metadata response to the `Metadata` object. Currently we primarily use the `Cluster` object for this, but we'll need another object for the leader epochs. Adding the leader epoch to `PartitionInfo` would have given this to us for free. Anyway, let's agree to cross that bridge when we get there. For now, I'll remove the field from `PartitionInfo`.

hachikuji committed:
Remove leader epoch field from PartitionInfo

hachikuji committed:
Address review comments

hachikuji commented:
@lindong28 Thanks for the detailed review. I think I've addressed all the pending comments and resolved the inconsistencies between the schemas documented in the KIP.

lindong28 commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -760,8 +761,9 @@ public void onComplete(Map<TopicPartition, OffsetAndMetadata> offsets, Exception
             }
-            offsetData.put(entry.getKey(), new OffsetCommitRequest.PartitionData(
-                    offsetAndMetadata.offset(), offsetAndMetadata.metadata()));
+            offsetData.put(entry.getKey(), new OffsetCommitRequest.PartitionData(offsetAndMetadata.offset(),
```
This line currently uses the deprecated constructor `PartitionData(long offset, long timestamp, String metadata)` which is probably not correct. Can you replace `offsetAndMetadata.leaderEpoch().orElse(RecordBatch.NO_PARTITION_LEADER_EPOCH)` with `offsetAndMetadata.leaderEpoch()`?

lindong28 commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -168,10 +169,12 @@ public Fetcher(LogContext logContext,
     private static class OffsetData {
         final long offset;
         final Long timestamp; //  null if the broker does not support returning timestamps
+        final Optional<Integer> leaderEpoch; // null if leader epoch is not known
```
nits: would it be a bit more accurate to say `Optional.empty() if leader epoch is not known`?

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -36,154 +35,166 @@
-
-    // FETCH_REQUEST_PARTITION_V5 added log_start_offset field - the earliest available offset of partition data that can be consumed.
-    private static final Schema FETCH_REQUEST_PARTITION_V5 = new Schema(
+    private static final Field.ComplexArray TOPICS = new Field.ComplexArray("responses",
```
It seems that previous field name `topics` is better than `responses`? `topics` is probably also more consistent with other schemas such as `ListOffsetRequest`.

There is a few other schemas that uses `responses`. I feel it may be good to keep the name unchanged if it is not too confusing. Otherwise, it seems better to change all of them to use `topics` consistently.

hachikuji replied:
Just to clarify, are you suggesting that we rename the field name "responses" to "topics"? I agree that would be clearer and more consistent with the other APIs, but you had some reluctance about changing field names before.

hachikuji replied:
Oh, my bad. This was a copy/paste error. The previous name was "topics." I will fix it.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -36,154 +35,166 @@
+                    "and enables the inclusion of the list of aborted transactions in the result, which allows " +
+                    "consumers to discard ABORTED transactional records");
+    private static final Field.Int32 SESSION_ID = new Field.Int32("session_id", "The fetch session ID");
+    private static final Field.Int32 SESSION_EPOCH = new Field.Int32("session_epoch", "The fetch session epoch");
```
I don't have a strong opinion to decide whether we can change the field name for existing schema versions. I agree this would not cause problem for deployment but on the other hand this could cause confusion for developers who look at the code to understand how we evolve the schema (including the field name). It would be good to have one more developer to comment on this. Could @ijuma comment on this?

ijuma replied:
The existing names are for humans to interpret, so as long as the new names can be easily inferred, I don't see a problem. We can also add an upgrade note if necessary.

hachikuji replied:
Yeah, I think it's probably adding an upgrade note if there was a semantic change, but I can add one if needed. I can also just revert the change, but I thought we may as well try to make the usage explicit given the presence of an additional epoch in the request.

lindong28 replied:
Sounds good.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -36,154 +35,166 @@
+    private static final Field.Int32 SESSION_EPOCH = new Field.Int32("session_epoch", "The fetch session epoch");
+
+    // topic level fields
+    private static final Field.ComplexArray PARTITIONS = new Field.ComplexArray("partition_responses",
```
It seems that the previous field name `partitions` is also clear and a bit more concise than `partition_responses`? `partitions` is probably also more consistent with other schemas such as `ListOffsetRequest`. There is a few other schemas that uses `partition_responses`. I feel it may be good to keep the name unchanged if it is not too confusing. Otherwise, it seems better to change all of them to use `partitions` consistently.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/ListOffsetRequest.java:
```diff
@@ -47,70 +45,93 @@
     private final IsolationLevel isolationLevel;
-    private final Map<TopicPartition, PartitionData> offsetData;
-    private final Map<TopicPartition, Long> partitionTimestamps;
+    private final Map<TopicPartition, PartitionData> partitionTimestamps;
```
nits: could we rename this variable (and related variables/methods) to e.g. `partitionDataMap`

lindong28 replied:
Hmm.. it looks like the change is not pushed? It is OK if you prefer `partitionTimestamps`.

lindong28 replied:
Never mind. This can be refactored in the followup patch.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetCommitRequest.java:
```diff
@@ -31,95 +29,112 @@
+    private static final Field PARTITIONS_V6 = PARTITIONS.withFields(
+            PARTITION_ID,
+            COMMITTED_OFFSET,
+            COMMITTED_LEADER_EPOCH,
```
The OffsetCommitRequest schema in KIP-320 has field `Timestamp`. It seems that this field has been removed from OffsetCommitRequest in the past. Can we update the KIP as appropriate?

hachikuji replied:
Good catch. Fixed the KIP.

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchResponse.java:
```diff
@@ -32,79 +31,94 @@
     private static final Schema OFFSET_FETCH_RESPONSE_V4 = OFFSET_FETCH_RESPONSE_V3;
 
+    // V5 adds the leader epoch to the committed offset
+    private static final Field PARTITIONS_V5 = PARTITIONS.withFields(
```
Could you update the KIP-320 so that its schema OffsetFetchResponse is consistent with the code here? The KIP-320 uses Timestamp in the schema and the position of LeaderEpoch is inconsistent.

lindong28 replied:
Thanks. I made a slight change in KIP-320 https://cwiki.apache.org/confluence/pages/diffpagesbyversion.action?pageId=87295403&selectedPageVersions=39&selectedPageVersions=40

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetsForLeaderEpochRequest.java:
```diff
@@ -99,23 +114,24 @@ public String toString() {
                 Struct partitionAndEpoch = (Struct) partitionAndEpochObj;
                 int partitionId = partitionAndEpoch.get(PARTITION_ID);
-                int epoch = partitionAndEpoch.getInt(LEADER_EPOCH);
+                int searchEpoch = partitionAndEpoch.get(LEADER_EPOCH);
```
nits: could `searchEpoch` be renamed to `leaderEpoch` for consistency?

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/TxnOffsetCommitResponse.java:
```diff
@@ -34,41 +33,49 @@
 public class TxnOffsetCommitResponse extends AbstractResponse {
-    private static final String TOPICS_KEY_NAME = "topics";
-    private static final String PARTITIONS_KEY_NAME = "partitions";
+    private static final Field.ComplexArray TOPICS = new Field.ComplexArray("responses",
```
Could we keep the existing name `topics`?

lindong28 commented on clients/src/main/java/org/apache/kafka/common/requests/TxnOffsetCommitResponse.java:
```diff
@@ -34,41 +33,49 @@
 
-    private static final Schema TXN_OFFSET_COMMIT_PARTITION_ERROR_RESPONSE_V0 = new Schema(
+    // topic level fields
+    private static final Field.ComplexArray PARTITIONS = new Field.ComplexArray("partition_responses",
```
Could we keep the existing name `partitions`?

lindong28 commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -63,7 +65,8 @@ class ReplicaFetcherThread(name: String,
   // Visible for testing
   private[server] val fetchRequestVersion: Short =
-    if (brokerConfig.interBrokerProtocolVersion >= KAFKA_2_0_IV1) 8
+    if (brokerConfig.interBrokerProtocolVersion >= KAFKA_2_0_IV1) 9
```
Should this be `KAFKA_2_1_IV1`? Same for `offsetForLeaderEpochRequestVersion` and `listOffsetRequestVersion`.

hachikuji replied:
Ouch. Good catch.

lindong28 replied:
Hmm... it seems that `KAFKA_2_0_IV1` is still used for `offsetForLeaderEpochRequestVersion` and `listOffsetRequestVersion`. Could you fix them as well?

lindong28 replied:
I made a commit to fix it :)

lindong28 commented on core/src/test/scala/unit/kafka/api/ApiVersionTest.scala:
```diff
@@ -80,7 +80,7 @@ class ApiVersionTest {
     assertEquals(KAFKA_2_0_IV1, ApiVersion("2.0-IV1"))
 
-    assertEquals(KAFKA_2_1_IV0, ApiVersion("2.1"))
+    assertEquals(KAFKA_2_1_IV1, ApiVersion("2.1"))
```
Can we also add assertion `assertEquals(KAFKA_2_1_IV1, ApiVersion("2.1-IV1"))`?

lindong28 reviewed (commented):
Thanks for the latest patch Jason. Looks good overall. Just a couple more comments below. If the new commit can avoid big refactor change and only fixes the issues mentioned below, I will probably commit the patch without going through the entire patch.

hachikuji committed:
Minor cleanup and fixes per review comments

hachikuji commented:
@lindong28 I really appreciate the help reviewing this. I've pushed an update to address your latest comments.

lindong28 committed:
Use KAFKA_2_1_IV1

lindong28 reviewed (approved):
@hachikuji Thanks much for the patch. I am very happy to help! I made a minor change in ReplicaFetcherThread.java to use `KAFKA_2_1_IV1` as appropriate. I will wait for test result before merging the PR.

hachikuji commented:
@lindong28 Ack on the changes to fix the inter-broker version checks. Thanks for the review!
